core/src/main/java/io/grpc/internal/MessageFramer.java

```diff
@@ -121,39 +113,52 @@ void setMessageCompression(boolean enable) {
    */
   public void writePayload(InputStream message) {
     verifyNotClosed();
+    boolean compressed = messageCompression && compressor != Codec.Identity.NONE;
+    int written = -1;
+    int messageLength = -2;
     try {
-      if (messageCompression && compressor != Codec.Identity.NONE) {
-        writeCompressed(message);
+      messageLength = getKnownLength(message);
+      if (messageLength != 0 && compressed) {
+        written = writeCompressed(message, messageLength);
       } else {
-        writeUncompressed(message);
+        written = writeUncompressed(message, messageLength);
       }
-    } catch (IOException ex) {
-      throw new RuntimeException(ex);
+    } catch (IOException e) {
+      // This should not be possible, since sink#deliverFrame doesn't throw.
+      throw Status.INTERNAL
+          .withDescription("Failed to frame message")
+          .withCause(e)
+          .asRuntimeException();
+    } catch (RuntimeException e) {
+      throw Status.INTERNAL
+          .withDescription("Failed to frame message")
+          .withCause(e)
+          .asRuntimeException();
+    }
+
+    if (messageLength != -1 && written != messageLength) {
+      String err = String.format("Message length inaccurate %s != %s", written, messageLength);
+      throw Status.INTERNAL.withDescription(err).asRuntimeException();
     }
   }
 
-  private void writeUncompressed(InputStream message) throws IOException {
-    int messageLength = getKnownLength(message);
+  private int writeUncompressed(InputStream message, int messageLength) throws IOException {
     if (messageLength != -1) {
-      writeKnownLength(message, messageLength, false);
-    } else {
-      BufferChainOutputStream bufferChain = new BufferChainOutputStream();
-      writeToOutputStream(message, bufferChain);
-      writeBufferChain(bufferChain, false);
+      return writeKnownLength(message, messageLength, false);
     }
+    BufferChainOutputStream bufferChain = new BufferChainOutputStream();
+    int written = writeToOutputStream(message, bufferChain);
+    writeBufferChain(bufferChain, false);
+    return written;
   }
 
-  private void writeCompressed(InputStream message) throws IOException {
+  private int writeCompressed(InputStream message, int messageLength) throws IOException {
     BufferChainOutputStream bufferChain = new BufferChainOutputStream();
-    // Why this doesn't use getKnownLength() idk, but let's just roll with it.
-    int messageLength = -1;
-    if (message instanceof KnownLength) {
-      messageLength = message.available();
-    }
 
     OutputStream compressingStream = compressor.compress(bufferChain);
+    int written;
     try {
-      long written = writeToOutputStream(message, compressingStream);
+      written = writeToOutputStream(message, compressingStream);
       if (messageLength != -1 && messageLength != written) {
         throw new RuntimeException("Message length was inaccurate");
```

Comment:
I had at one point tried to merge write compressed and write uncompressed so as to not have the duplicate exception, but that turned out to be pretty ugly.  I reverted back to this version, but I guess I forgot to remove this.